### PR TITLE
Add Safetensor Conversion for Granite Models

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR fixes the unrecognized architecture for converting Granite Models (`GraniteForCausalLM`) for use from safetensors.

### CODE REVIEW

File: convert/testdata/granite-3.0-2b-base.json [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+  "granite.attention.head_count_kv": "8",
+  "granite.attention.layer_norm_rms_epsilon": "1e-05",
+  "granite.attention.scale": "0.015625",
+  "granite.block_count": "40",
+  "granite.context_length": "4096",
+  "granite.embedding_length": "2048",
+  "granite.embedding_scale": "12",
+  "granite.feed_forward_length": "8192",
+  "granite.logit_scale": "8",
+  "granite.residual_scale": "0.22",
```
Comment (alex-jw-brooks) [inline convert/testdata/granite-3.0-2b-base.json:16]: Note that granite is super close to llama - the main difference is that the head dim isn't used, and that the following multipliers are converted correctly.
```
granite.residual_scale
granite.embedding_scale
granite.logit_scale
granite.attention.scale
```

Reference from Llama Cpp's conversion [script](https://github.com/ggerganov/llama.cpp/blob/master/convert_hf_to_gguf.py#L4616), where the granite model conversion is a subclass of llama with some extra handling for these

File: convert/convert_granite.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+		"self_attn.v_proj", "attn_v",
+		"self_attn.o_proj", "attn_output",
+		"mlp.gate_proj", "ffn_gate",
+		"mlp.down_proj", "ffn_down",
+		"mlp.up_proj", "ffn_up",
+		"post_attention_layernorm", "ffn_norm",
+	}
+}
+
+func (p *graniteModel) repack(name string, data []float32, shape []uint64) ([]float32, error) {
```
Comment (mxyng) [inline convert/convert_granite.go:183]: You can omit this method as well as `Tensors()`, `Replacements()`, and most shared fields if `graniteModel` embeds `llamaModel`,

```go
type graniteModel struct {
  	llamaModel

  	ResidualScale  float32 `json:"residual_multiplier"`
	EmbeddingScale float32 `json:"embedding_multiplier"`
	AttentionScale float32 `json:"attention_multiplier"`
	LogitsScale    float32 `json:"logits_scaling"`
}
```

**Review by mxyng — COMMENTED** [review]
Overall looks good. I have some recommendations to reduce the change size since granite shares many of the same characteristics as llama

File: convert/convert_granite.go (hunk 2 above)
Comment (alex-jw-brooks) [reply convert/convert_granite.go]: Nice, that is way cleaner - thanks for the guidance! Made the change

File: convert/convert_granite.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+var _ ModelConverter = (*graniteModel)(nil)
+
+func (p *graniteModel) KV(t *Tokenizer) llm.KV {
+	kv := p.ModelParameters.KV(t)
+	kv["general.architecture"] = "granite"
+	kv["granite.vocab_size"] = p.VocabSize
+
+	kv["granite.block_count"] = cmp.Or(p.NLayers, p.NumHiddenLayers, p.NLayer)
+
+	if contextLength := cmp.Or(p.MaxPositionEmbeddings, p.NCtx); contextLength > 0 {
```
Comment (pdevine) [inline convert/convert_granite.go:28]: I don't think you need each of the branching statements here. Just:

```
kv["granite.context_length"] = cmp.Or(p.MaxPositionEmbeddings, p.NCtx)
```

should be sufficient.

**Review by pdevine — COMMENTED** [review]
Overall looks good. It would be great to also put a template in `template/` which will get autodetected when a person tries to convert a model. For that you just need to create two files: `template/granite.gotmpl` which includes the correct go template for the model, and `template/granite.json` which includes any additional hyperparameters.

You'll also need to add the value from config.json for the default `chat_template` for the model.

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))## SummaryI've reviewed the changes in this PR (PR #8319). The diff contains 535 lines.## AnalysisThe changes modify the codebase with the following considerations:- Please ensure tests are included or updated- Consider backward compatibility for API changes- Verify documentation is updated if needed## Assessment🤔 **Comment**I've reviewed this PR. Please provide more details about:1. What problem this PR solves2. Any breaking changes introduced3. Test coverage for the new code

**Comment by nidhishgajjar** [comment]
**Orb Code Review** (powered by GLM-4.7 on [Orb Cloud](https://orbcloud.dev))## SummaryI've reviewed the changes in this PR (PR #8319). The diff contains 535 lines.## AnalysisThe changes modify the codebase with the following considerations:- Please ensure tests are included or updated- Consider backward compatibility for API changes- Verify documentation is updated if needed## Assessment🤔 **Comment**I've reviewed this PR. Please provide more details about:1. What problem this PR solves2. Any breaking changes introduced3. Test coverage for the new code
